worlds/cv64/__init__.py

```diff
@@ -0,0 +1,313 @@
+        ["Liquid Cat"]
+    )]
+
+
+class CV64World(World):
+    """
+    Castlevania for the Nintendo 64 is the first 3D game in the franchise. As either whip-wielding Belmont descendant
+    Reinhardt Schneider or powerful sorceress Carrie Fernandez, brave many terrifying traps and foes as you make your
+    way to Dracula's chamber and stop his rule of terror.
+    """
+    game = "Castlevania 64"
+    item_name_groups = {
+        "Bomb": {iname.magical_nitro, iname.mandragora},
+        "Ingredient": {iname.magical_nitro, iname.mandragora},
+    }
+    location_name_groups = {stage: set(get_locations_from_stage(stage)) for stage in vanilla_stage_order}
+    options_dataclass = CV64Options
+    options: CV64Options
+    settings: typing.ClassVar[CV64Settings]
+    topology_present = True
+    data_version = 1
+
+    item_name_to_id = get_item_names_to_ids()
+    location_name_to_id = get_location_names_to_ids()
+
+    active_stage_exits: typing.Dict[str, typing.Dict]
+    active_stage_list: typing.List[str]
+    active_warp_list: typing.List[str]
+    reinhardt_stages: bool
+    carrie_stages: bool
+    branching_stages: bool
+    starting_stage: str
+
+    total_s1s: int
+    s1s_per_warp: int
+    total_s2s: int
+    required_s2s: int
+    drac_condition: int
+
+    auth: bytearray
+
+    web = CV64Web()
+
+    @classmethod
+    def stage_assert_generate(cls, multiworld: MultiWorld) -> None:
+        rom_file = get_base_rom_path()
+        if not os.path.exists(rom_file):
+            raise FileNotFoundError(rom_file)
+
+    def generate_early(self) -> None:
+        # Generate the player's unique authentication
+        self.auth = bytearray(self.multiworld.random.getrandbits(8) for _ in range(16))
+
+        self.total_s1s = self.options.total_special1s.value
+        self.s1s_per_warp = self.options.special1s_per_warp.value
+        self.drac_condition = self.options.draculas_condition.value
+
+        # If there are more S1s needed to unlock the whole warp menu than there are S1s in total, drop S1s per warp to
+        # something manageable.
+        while self.s1s_per_warp * 7 > self.total_s1s:
+            self.s1s_per_warp -= 1
+
+        # Set the total and required Special2s to 1 if the drac condition is the Crystal, to the specified YAML numbers
+        # if it's Specials, or to 0 if it's None or Bosses. The boss totals will be figured out later.
+        if self.drac_condition == DraculasCondition.option_crystal:
+            self.total_s2s = 1
+            self.required_s2s = 1
+        elif self.drac_condition == DraculasCondition.option_specials:
+            self.total_s2s = self.options.total_special2s.value
+            self.required_s2s = int(self.options.percent_special2s_required.value / 100 * self.total_s2s)
+        else:
+            self.total_s2s = 0
+            self.required_s2s = 0
+
+        # Enable/disable character stages and branching paths accordingly
+        self.reinhardt_stages = True
```

Comment:
per my previous comment on not assigning a ton of attributes in the constructor, these are not mutable so can just be assigned in the class definition. The only objects you need to worry about assigning in the class definition are those which *are* [mutable](https://docs.python.org/3/library/stdtypes.html#mutable-sequence-types) (most commonly lists or dictionaries), so specifically your active_exits/lists. this is because if you have a mutable object assigned in the class definition, modifying the instance attributes will find the class attribute and just modify that instead of making its own. for non-immutable types this isn't really viable so it has to have its own attribute instead of stealing the one owned by the class. All the other attributes you have such as these booleans can just be set in the class definition and then changed later. ultimately it really isn't that big of a deal but it's worth understanding when it's appropriate.

Reply:
I encouraged moving all initialization into the constructor in my review. I suppose moving immutable stuff into a constructor isn't entirely necessary and uses up the small amount of extra memory that a bool or whatever takes up, but otherwise I prefer the consistency of initializing attributes that are relevant only to _instances_ of a class in `__init__`, since it feels more intentional/proper.

I'm not an experienced python dev, so I'm happy to defer here. Just giving the context of my opinion since I gave slightly contradictory advice.

Reply:
since python isn't a strongly typed language both are valid, and initializing all of them in the constructor is generally best practice if you *aren't* using type hints (which is a bad practice in and of itself imo). the point of these peer reviews is to try and encourage best practices though and when you have type hints available, typing everything and defining those that don't need to be instance specific in the class definition; and those that do need to be instance specific in the constructor is [best practice](https://peps.python.org/pep-0526/#class-and-instance-variable-annotations). the only reason i try to lead people away from using the constructor specifically in world definitions is because doing anything other than very basic default setting is dangerous in the constructor for our uses. generate_early is the earliest method where you can realistically do things with your world, and *most* worlds need to define generate_early to do some stuff anyways so it just makes the most sense to instantiate stuff then, or even better as you need it.

the bit where it's actually important is really just style consistency; which in our case can be divided down to the world level, but i'd still like to try and encourage core and worlds have as much stylistic consistency as possible. i definitely still want to stress that this really isn't that big of a deal though.

Reply:
Ok. I *think* I have it "correct" this time.

Reply:
yep looks good 👍